File: src/Common in 3 Sorted Arrays/Code/solution.py
class Solution:
    def commonElements(self, a, b, c):
        ans = []
        i = j = k = 0

        while i < len(a) and j < len(b) and k < len(c):
            x, y, z = a[i], b[j], c[k]

            # Case 1: all three are equal => common element found
            if x == y == z:
                # Avoid duplicates in the answer
                if not ans or ans[-1] != x:
                    ans.append(x)
                i += 1
                j += 1
                k += 1
            # Case 2: move the pointer(s) having the smallest value
            else:
                mn = min(x, y, z)
                if x == mn:
                    i += 1
                if y == mn:
                    j += 1
                if z == mn:
                    k += 1

        return ans

File: src/Common in 3 Sorted Arrays/Code/solution.cpp
class Solution
{
public:
    vector<int> commonElements(vector<int> &a, vector<int> &b, vector<int> &c)
    {
        vector<int> ans;
        int i = 0, j = 0, k = 0;

        while (i < (int)a.size() && j < (int)b.size() && k < (int)c.size())
        {
            int x = a[i], y = b[j], z = c[k];

            // Case 1: all three are equal => common element found
            if (x == y && y == z)
            {
                // Avoid duplicates in the answer
                if (ans.empty() || ans.back() != x)
                {
                    ans.push_back(x);
                }
                i++;
                j++;
                k++;
            }
            // Case 2: move the pointer(s) having the smallest value
            else
            {
                int mn = min(x, min(y, z));
                if (x == mn)
                    i++;
                if (y == mn)
                    j++;
                if (z == mn)
                    k++;
            }
        }

        return ans;
    }
};